### DISCUSSION TITLE
How to build and use Bluetooth enabled variant of unix port

### DISCUSSION THREAD

**Discussion by mogenson** [micropython/micropython#10234] [Unix & Windows]
Hello, how do I build and use the unix port with Bluetooth enabled?

I see there's some nimble and btstack source files in the unix port directory, some commits by @jimmo prototyping initial Bluetooth support on unix, a few mentions of using the unix port to test BLE, and an open PR for a Windows [micropython-ble](https://github.com/micropython/micropython/pull/7781/files#diff-a0f75b1837f9125378b005550ba49bb99b91a67246230834e59a73a0833d0c4c) variant by @andrewleech .

What configs/defines need to be enabled to attach the micropython unix interpreter to a BT controller over UART or USB HCI? How is this setup at runtime? Are both the btstack and nimble stacks supported equally?

Finally, would a PR be welcome to create a BLE variant for the unix port and add some getting started documentation or command line help?

**Comment by andrewleech** [comment]
Hi, I've actually written unix ble variant's a couple of times but always found a reason to not push/pr it... usually because I hadn't made the time to write enough supporting docs to make it useful :-/  

Micropython supports two different stacks on unix, btstack and nimble. btstack is free for open source use, but needs to be licenced for commercial use. Nimble is free for anyone to use.

btstack in theory supports generic usb bluetooth dongles (build with `MICROPY_BLUETOOTH_BTSTACK_USB=1 MICROPY_BLUETOOTH_BTSTACK_H4=0`) however I never had any luck making this work.

Both stacks support "HCI over (usb) Uart" where you have a bluetooth micro set up to basically proxy the HCI commands from usb/uart to the ble controller. 

I put together firmware for the Nordic nrf52840 dongle to do this: https://gitlab.com/alelec/nrf52810_zephyr_hci_controller a build of which can be downloaded here: https://gitlab.com/alelec/zephyr_bluetooth_hci_usb_uart/-/jobs/793847460/artifacts/download
Download the firmware pack, unzip it. Plug nrf dongle into pc and press the "90 degree / sideways switch" near the end and its led will start "pulsing" 
![image](https://user-images.githubusercontent.com/3318786/207984523-b071491a-a6c5-4588-959b-6cbda500eb30.png)
Check which uart / com port it comes up as and use nrfutil to flash it
`nrfutil.exe dfu serial -pkg nrf52840_zephyr_ble_controller_4b594b1b\nrf52840_zephyr_hci_usb_uart.zip -p COM3`

@jimmo has since cleaned up my code for this and submitted it as an official zephyr sample application so it can now be build from there: https://github.com/zephyrproject-rtos/zephyr/pull/50453

You can also use an stm32wb55 nucleo dongle with Transparant VCP firmware aka https://github.com/micropython/micropython/pull/9046

If you're using a Windows PC and want to test with WSL, you can connect your USB devices through to WSL with https://gitlab.com/alelec/wsl-usb-gui

To build micropython unix port with BLE:
```
cd micropython

# To build with btstack
make -C ports/unix MICROPY_PY_BLUETOOTH=1 MICROPY_BLUETOOTH_BTSTACK=1 submodules
# btstack with usb/uart dongle support
make -C ports/unix MICROPY_PY_BLUETOOTH=1 MICROPY_BLUETOOTH_BTSTACK=1 MICROPY_BLUETOOTH_BTSTACK_USB=0 MICROPY_BLUETOOTH_BTSTACK_H4=1

# nimble stack
make -C ports/unix MICROPY_PY_BLUETOOTH=1 MICROPY_BLUETOOTH_NIMBLE=1 submodules
make -C ports/unix MICROPY_PY_BLUETOOTH=1 MICROPY_BLUETOOTH_NIMBLE=1

# run micropython
#  - check which com port your dongle connects as, micropython defaults to /dev/ttyUSB0 but can be changed with MICROPYBTUART env

MICROPYBTUART=/dev/ttyACM0 ports/unix/build-standard/micropython
```
```
MicroPython v1.19.1-773-g988b6e2da on 2022-12-16; linux [GCC 9.4.0] version
Use Ctrl-D to exit, Ctrl-E for paste mode
>>> from bluetooth import BLE
>>> BLE().active(1)
True
>>>

**Comment by jimmo** [comment]
hi @mogenson 

Every time I go to make a PR to add BLE variants for Unix I end up going down a rabbit hole of trying to build "sub variant" support -- i.e. "standard with nimble" or "standard with btstack".

I do regularly use the Unix port with BLE support though for testing and development of BLE on MicroPython. The exact build instructions @andrewleech posted above should work.

I mostly use the nRF-dongle-running-zephyr (using the example firmware that Andrew linked to -- referenced in the README here https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/bluetooth/hci_uart#using-a-usb-cdc-acm-uart). This works with NimBLE and btstack. I also occasionally use the WB55 transparent mode that Andrew mentioned.

For btstack in usb mode, I also have a USB dongle based on a BCM20702.

It would also be possible in theory to use your host BLE (via bluez btproxy), but the MicroPython hci code would need to be updated to be able to talk to the btproxy socket.

If you want to use aioble on the Unix port, then you also need

```
#define MICROPY_PY_USELECT_POSIX (0)
#define MICROPY_PY_USELECT (1)
```

in your mpconfigvariant.h.

I'm not sure adding a variant is necessary, but we would definitely welcome any documentation updates!

**Comment by mogenson** [comment]
Thank you for the great information! I'll give this a try!
